### PR TITLE
host-dma: only do DMA L1 exit at post copy

### PR DESCRIPTION
**Pull request by keyonjie**
Only set DMA L1 Exit at post copy to reduce the power consumption, this
will benefit the KWD scenario as kept the HDA host DMA at L1 Exit even
no real data is being copied will stop the system to enter s0ix low
power stage.

Signed-off-by: Keyon Jie <yang.jie@linux.intel.com>

### CODE REVIEW

File: src/audio/host.c [hunk 1]
```diff
@@ -383,6 +383,13 @@ static int host_copy_normal(struct comp_dev *dev)
 		return ret;
 	}
 
+	if (hd->chan->status != COMP_STATE_ACTIVE) {
+		ret = dma_start(hd->chan);
```
Comment (ranj063) [inline src/audio/host.c:387]: @keyonjie the commit message doesnt explain why this change is needed? Isnt the dma started already during the START trigger?

**Comment by ranj063** [comment]
> This also looks like a revert of @tlauda work on the same topic. What's the background here ?

@lgirdwood the rationale for the change is that the even without the keyword detection pipeline not copying anything to the host memory before the keyword is detected, the DDR was unable to enter self-refresh thereby preventing the system from entering S0ix. We verified this by starting the DMA before the keyword only after the keyword is detected and saw no issues with system S0ix.

**Comment by aiChaoSONG** [comment]
SOFCI TEST

File: src/audio/host.c (hunk 1 above)
Comment (keyonjie) [reply src/audio/host.c:387]: oops, this was kept by mistake from the old version, it is not needed anymore, let me update it.

**Comment by keyonjie** [comment]
> This also looks like a revert of @tlauda work on the same topic. What's the background here ?

Yes, at least partially. I talked with @tlauda about this yesterday. Since we have changed most of our pipelines to timer driver already, and for timer driver scenarios, the difference with today's code is that we will do an DMA L1 exit at each scheduling period (20ms for KWD pipelines), even though there is no any data copied during the period, this will lead to the host buffer (DDR) being refreshed every period and the whole system can't stay at deep power saving stage for long time.

With the change applied, we will do the unified DMA L1 exit handling for both timer driven and DMA irq driver pipelines, that is, doing the DMA L1 exit at each hda_dma_post_copy() where means we do have already copied some "bytes" and the DDR content refresh is really needed.

So the change looks safe to me, @tlauda please help to give your comments about it, I would require for more test (including other playback/capture pipelines, and on platforms other than TGL).

**Comment by keyonjie** [comment]
> @keyonjie Could you submit changes to TGL release branch first tgl-007-drop-stable ?

Sure, let me do that to unblock the release.

**Comment by tlauda** [comment]
@keyonjie @lgirdwood The direction of this change is good - there is no point in exiting L1 without copying any data. However, the change itself reverts my previous optimization. For timer driven systems we don't want to exit L1 for every host DMA, because it increases host component's MCPS consumption significantly. That's why it was moved to one common place (after all LL tasks are done). Before merging to master you will need to somehow refactor this patch.

**Comment by keyonjie** [comment]
> @keyonjie @lgirdwood The direction of this change is good - there is no point in exiting L1 without copying any data. However, the change itself reverts my previous optimization. For timer driven systems we don't want to exit L1 for every host DMA, 

Thanks for the information @tlauda .
Do you mean that "for every host DMA channel"? I am not quite clear about how that worked, for example, if we run 2 simultaneous pipelines, do you mean with today's code it will do only one L1 exit at each period pipeline_copy() done?

> because it increases host component's MCPS consumption significantly. That's why it was moved to one common place (after all LL tasks are done). Before merging to master you will need to somehow refactor this patch.

Can you help to share more info about this MCPS consumption?

**Comment by tlauda** [comment]
> > @keyonjie @lgirdwood The direction of this change is good - there is no point in exiting L1 without copying any data. However, the change itself reverts my previous optimization. For timer driven systems we don't want to exit L1 for every host DMA,
> 
> Thanks for the information @tlauda .
> Do you mean that "for every host DMA channel"? I am not quite clear about how that worked, for example, if we run 2 simultaneous pipelines, do you mean with today's code it will do only one L1 exit at each period pipeline_copy() done?

Yes, precisely, only one L1 exit for all host DMA channels. You can check in the code, that we are registering pm_runtime_get and put methods on LL_PRE and POST notification events, which are called in low latency scheduler only once per all tasks.

> 
> > because it increases host component's MCPS consumption significantly. That's why it was moved to one common place (after all LL tasks are done). Before merging to master you will need to somehow refactor this patch.
> 
> Can you help to share more info about this MCPS consumption?

L1 exit is basically a static wait for predefined number of cycles. I've done some experiments in the past and it increased host copy execution even two times.

File: src/drivers/intel/hda/hda-dma.c [hunk 2]
```diff
@@ -221,18 +221,6 @@ static void hda_dma_get_dbg_vals(struct dma_chan_data *chan,
 #define hda_dma_ptr_trace(...)
```
Comment (plbossart) [inline src/drivers/intel/hda/hda-dma.c:221]: @keyonjie it took me several minutes to understand the commit message, and I think you are mixing L1 entry/exit.

suggested edit:

Exiting the DMA L1 state, e.g. on a timer tick when there is no data to copy, prevents the system from staying in S0ix power states. To extend S0ix residency, this L1 exit needs to happen only when copying data and the L1 mode can be re-entered in the post-copy stage.

Comment (keyonjie) [reply src/drivers/intel/hda/hda-dma.c:221]: Thanks @plbossart . We don't have real L1 Entry:
In L1 Entry, we just increase the ref count host_dma_l1_sref.
In L1 Exit, we decrease the host_dma_l1_sref, and toggle the L1 Exit bit to force a L1 exit (flushing data from L1 Cache to DDR) if host_dma_l1_sref is 0.

So, the L1 exit does not happen when copying data, it should happen after those data are copied (that is, in the post-copy stage).

Comment (keyonjie) [reply src/drivers/intel/hda/hda-dma.c:221]: @plbossart I take your suggestion and do a minor change according to what I explained above, updated now.

**Comment by lgirdwood** [comment]
@keyonjie something funny with your PR, can you make sure rimage submodule is up to date and force push again, it's showing an rimage delta.

File: src/drivers/intel/hda/hda-dma.c [hunk 3]
```diff
@@ -309,9 +296,8 @@ static void hda_dma_post_copy(struct dma_chan_data *chan, int bytes)
 		 */
 		hda_dma_inc_fp(chan, bytes);
 
-		/* Force Host DMA to exit L1 if scheduled on DMA */
-		if (!hda_chan->irq_disabled)
-			pm_runtime_put(PM_RUNTIME_HOST_DMA_L1, 0);
+		/* Force Host DMA to exit L1 */
+		pm_runtime_put(PM_RUNTIME_HOST_DMA_L1, 0);
```
Comment (plbossart) [inline src/drivers/intel/hda/hda-dma.c:300]: I must admit I am quite confused...
I can see that for the LMEM-HMEM transfer you can do the L1 exit after copying the data, e.g. to force data to be transfered. 

however, if you do a HMEM-LMEM transfer, does it really make sense to exit L1 *after* copying the data?

File: src/drivers/intel/hda/hda-dma.c [hunk 4]
```diff
@@ -221,18 +221,6 @@ static void hda_dma_get_dbg_vals(struct dma_chan_data *chan,
 #define hda_dma_ptr_trace(...)
 #endif
 
-static void hda_dma_l1_entry_notify(void *arg, enum notify_id type, void *data)
```
Comment (plbossart) [inline src/drivers/intel/hda/hda-dma.c:224]: comment on commit message:

and the L1 mode will be re-entered after the L1 exit
done by toggling the SHIM_SVCFG_FORCE_L1_EXIT bit.

there is nothing in this patch that actually refers to this bit, and there is no explanation on when the L1 mode is re-entered, in other words you are saying both too much and too little at the same time.

**Comment by plbossart** [comment]
> @keyonjie something funny with your PR, can you make sure rimage submodule is up to date and force push again, it's showing an rimage delta.

I also get these rimage submodule deltas, annoying as heck. I can't figure out how to realign.

**Comment by keyonjie** [comment]
> This patch reverts good optimization that reduced multiple prevent-L1-entry loops (one per pipeline) to a single one run by ll-schedule systick. The original optimized code unnecessarily prevents l1-entry if registered host component is running while there is no data copied (case when kpbuffer blocks the output until a keyphrase is detected).
> Since there is another solution discussed and developed, with host component selectively suspended by the ipc trigger when there is no data to transfer, L1 should not be affected. And the original optimized solution does not need to be reverted.

Makes sense, that is what Tomek concerned also, let me closing this and open a new PR with new solution.
